File: backend/src/falkordb/graph-delta.ts
import type { DecisionEvent, GraphEdge } from '../schemas/decision.schema';
import { cypherRelType } from './cypher-delta';

export type GraphSnapshot = {
  nodeIds: Set<string>;
  edgeKeys: Set<string>;
};

export type EdgeOperation = {
  source: string;
  target: string;
  relType: string;
  edge: GraphEdge;
  targetId: string;
  isReverse: boolean;
};

/** Canonical directed edge identity for dedup. */
export function edgeKey(source: string, relType: string, target: string): string {
  return `${source}|${relType}|${target}`;
}

export function snapshotFromRows(
  nodes: { id: string }[],
  edges: { source: string; target: string; type: string }[],
): GraphSnapshot {
  return {
    nodeIds: new Set(nodes.map((n) => n.id)),
    edgeKeys: new Set(edges.map((e) => edgeKey(e.source, e.type, e.target))),
  };
}

function targetIds(edge: GraphEdge): string[] {
  if (edge.target) return [edge.target];
  return [...(edge.targets ?? [])];
}

/** Expand payload edges into directed MERGE operations (forward + optional reverse). */
export function expandEdgeOperations(edges: GraphEdge[]): EdgeOperation[] {
  const ops: EdgeOperation[] = [];
  const seen = new Set<string>();

  for (const edge of edges) {
    if (edge.metadata?.direction === 'reverse') continue;

    const relType = cypherRelType(edge);
    const targets = targetIds(edge);

    for (const targetId of targets) {
      const forwardKey = edgeKey(edge.source, relType, targetId);
      if (!seen.has(forwardKey)) {
        seen.add(forwardKey);
        ops.push({
          source: edge.source,
          target: targetId,
          relType,
          edge,
          targetId,
          isReverse: false,
        });
      }

      if (edge.direction === 'bi-directional') {
        const reverseKey = edgeKey(targetId, relType, edge.source);
        if (!seen.has(reverseKey)) {
          seen.add(reverseKey);
          ops.push({
            source: targetId,
            target: edge.source,
            relType,
            edge,
            targetId,
            isReverse: true,
          });
        }
      }
    }
  }

  return ops;
}

/** JSON vs graph: only nodes/edge ops not already present. */
export function computeGraphDelta(
  snapshot: GraphSnapshot,
  payload: DecisionEvent,
): {
  nodes: DecisionEvent['nodes'];
  edgeOps: EdgeOperation[];
  stats: {
    nodesSkipped: number;
    edgeOpsSkipped: number;
    nodesNew: number;
    edgeOpsNew: number;
  };
} {
  const nodes = payload.nodes.filter((n) => {
    if (snapshot.nodeIds.has(n.node_id)) return false;
    return true;
  });

  const allOps = expandEdgeOperations(payload.edges);
  const edgeOps = allOps.filter((op) => {
    const key = edgeKey(op.source, op.relType, op.target);
    return !snapshot.edgeKeys.has(key);
  });

  return {
    nodes,
    edgeOps,
    stats: {
      nodesSkipped: payload.nodes.length - nodes.length,
      edgeOpsSkipped: allOps.length - edgeOps.length,
      nodesNew: nodes.length,
      edgeOpsNew: edgeOps.length,
    },
  };
}

File: backend/app/falkordb/graph_delta.py
from dataclasses import dataclass
from typing import Any

from app.schemas.decision import DecisionEvent, GraphEdge


@dataclass
class EdgeOperation:
    source: str
    target: str
    rel_type: str
    edge: GraphEdge
    target_id: str
    is_reverse: bool


def edge_key(source: str, rel_type: str, target: str) -> str:
    return f"{source}|{rel_type}|{target}"


def cypher_rel_type(edge: GraphEdge) -> str:
    if edge.Action:
        return edge.Action.replace(" ", "_").upper()
    rel = edge.relationship_type or str(edge.metadata.get("relationship_type", "RELATED_TO"))
    cleaned = "".join(c if c.isalnum() or c == "_" else "_" for c in rel.upper())
    return cleaned or "RELATED_TO"


def snapshot_from_rows(
    nodes: list[dict[str, str]],
    edges: list[dict[str, str]],
) -> dict[str, set[str]]:
    return {
        "nodeIds": {n["id"] for n in nodes},
        "edgeKeys": {edge_key(e["source"], e["type"], e["target"]) for e in edges},
    }


def _target_ids(edge: GraphEdge) -> list[str]:
    if edge.target:
        return [edge.target]
    return list(edge.targets or [])


def expand_edge_operations(edges: list[GraphEdge]) -> list[EdgeOperation]:
    ops: list[EdgeOperation] = []
    seen: set[str] = set()

    for edge in edges:
        if edge.metadata.get("direction") == "reverse":
            continue
        rel_type = cypher_rel_type(edge)
        for target_id in _target_ids(edge):
            forward_key = edge_key(edge.source, rel_type, target_id)
            if forward_key not in seen:
                seen.add(forward_key)
                ops.append(
                    EdgeOperation(
                        source=edge.source,
                        target=target_id,
                        rel_type=rel_type,
                        edge=edge,
                        target_id=target_id,
                        is_reverse=False,
                    )
                )
            if edge.direction == "bi-directional":
                reverse_key = edge_key(target_id, rel_type, edge.source)
                if reverse_key not in seen:
                    seen.add(reverse_key)
                    ops.append(
                        EdgeOperation(
                            source=target_id,
                            target=edge.source,
                            rel_type=rel_type,
                            edge=edge,
                            target_id=target_id,
                            is_reverse=True,
                        )
                    )
    return ops


def compute_graph_delta(snapshot: dict[str, set[str]], payload: DecisionEvent) -> dict[str, Any]:
    nodes = [n for n in payload.nodes if n.node_id not in snapshot["nodeIds"]]
    all_ops = expand_edge_operations(payload.edges)
    edge_ops = [
        op
        for op in all_ops
        if edge_key(op.source, op.rel_type, op.target) not in snapshot["edgeKeys"]
    ]
    return {
        "nodes": nodes,
        "edgeOps": edge_ops,
        "stats": {
            "nodesSkipped": len(payload.nodes) - len(nodes),
            "edgeOpsSkipped": len(all_ops) - len(edge_ops),
            "nodesNew": len(nodes),
            "edgeOpsNew": len(edge_ops),
        },
    }
